torch/cuda/__init__.py

```diff
@@ -322,40 +323,82 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
     return torch._C._cuda_canDeviceAccessPeer(device, peer_device)
 
 
-@contextlib.contextmanager
-def stream(stream):
+class StreamContext(object):
     r"""Context-manager that selects a given stream.
 
     All CUDA kernels queued within its context will be enqueued on a selected
     stream.
 
     Args:
+        Stream (Stream): selected stream. This manager is a no-op if it's
+            ``None``.
+    .. note:: Streams are per-device.
+    """
+    cur_stream : Optional['torch.cuda.Stream']
+
+    def __init__(self, stream: Optional['torch.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        # Initialize the below streams to default stream on the current device
+        self.src_prev_stream = torch.cuda.default_stream(None)
+        self.dst_prev_stream = torch.cuda.default_stream(None)
```

Comment:
You don't need to initialize these if stream is None, because the context manager is a noop in that case.

Reply:
Agreed. TorchScript complains if we do something like this:
```
if self.stream is not None:
        self.src_prev_stream = torch.cuda.default_stream(None)
        self.dst_prev_stream = torch.cuda.default_stream(None)

RuntimeError: 
First assignment cannot be in a control-flow block. Initialize the field at the top level first:
```